File: app/src/main/java/piuk/blockchain/android/ui/chooser/AccountChooserActivity.java
package piuk.blockchain.android.ui.chooser;

import android.content.Intent;
import android.databinding.DataBindingUtil;
import android.os.Bundle;
import android.support.annotation.Nullable;
import android.support.v4.app.Fragment;
import android.support.v7.widget.LinearLayoutManager;
import android.view.View;

import com.fasterxml.jackson.core.JsonProcessingException;
import com.fasterxml.jackson.databind.ObjectMapper;

import java.util.List;

import javax.inject.Inject;

import piuk.blockchain.android.BuildConfig;
import piuk.blockchain.android.R;
import piuk.blockchain.android.data.contacts.models.PaymentRequestType;
import piuk.blockchain.android.databinding.ActivityAccountChooserBinding;
import piuk.blockchain.android.injection.Injector;
import piuk.blockchain.android.ui.account.ItemAccount;
import piuk.blockchain.android.ui.base.BaseMvpActivity;


public class AccountChooserActivity extends BaseMvpActivity<AccountChooserView, AccountChooserPresenter>
        implements AccountChooserView {

    public static final String EXTRA_REQUEST_TYPE = "request_type";
    public static final String EXTRA_REQUEST_CODE = "request_code";
    public static final String EXTRA_SELECTED_ITEM = "selected_object";
    public static final String EXTRA_SELECTED_OBJECT_TYPE = "selected_object_type";
    public static final String EXTRA_ACTIVITY_TITLE = "activity_title";
    public static final int REQUEST_CODE_CHOOSE_RECEIVING_ACCOUNT_FROM_RECEIVE = 216;
    public static final int REQUEST_CODE_CHOOSE_RECEIVING_ACCOUNT_FROM_SEND = 217;
    public static final int REQUEST_CODE_CHOOSE_SENDING_ACCOUNT_FROM_SEND = 218;
    public static final int REQUEST_CODE_CHOOSE_CONTACT = 219;

    @Inject AccountChooserPresenter accountChooserPresenter;
    private ActivityAccountChooserBinding binding;
    private PaymentRequestType paymentRequestType;

    {
        Injector.getInstance().getPresenterComponent().inject(this);
    }

    @Override
    protected void onCreate(@Nullable Bundle savedInstanceState) {
        super.onCreate(savedInstanceState);
        binding = DataBindingUtil.setContentView(this, R.layout.activity_account_chooser);

        Intent intent = getIntent();
        if (!intent.hasExtra(EXTRA_REQUEST_TYPE) || intent.getSerializableExtra(EXTRA_REQUEST_TYPE) == null) {
            throw new AssertionError("Request type must be passed to AccountChooserActivity");
        } else {
            paymentRequestType = (PaymentRequestType) intent.getSerializableExtra(EXTRA_REQUEST_TYPE);
            int requestCode = getIntent().getIntExtra(EXTRA_REQUEST_CODE, -1);

            String title = getIntent().getStringExtra(EXTRA_ACTIVITY_TITLE);
            if (title == null) {
                throw new AssertionError("Title string must be passed to AccountChooserActivity");
            } else {
                binding.toolbar.toolbarGeneral.setTitle(title);
            }

            setSupportActionBar(binding.toolbar.toolbarGeneral);
            if (getSupportActionBar() != null) {
                getSupportActionBar().setDisplayHomeAsUpEnabled(true);
            }
        }

        onViewReady();
    }

    @Override
    public void showNoContacts() {
        binding.recyclerview.setVisibility(View.GONE);
        binding.layoutNoContacts.setVisibility(View.VISIBLE);
    }

    @Override
    public boolean isContactsEnabled() {
        return BuildConfig.CONTACTS_ENABLED;
    }

    @Override
    public boolean onSupportNavigateUp() {
        setResult(RESULT_CANCELED);
        onBackPressed();
        return super.onSupportNavigateUp();
    }

    @Override
    public PaymentRequestType getPaymentRequestType() {
        return paymentRequestType;
    }

    @Override
    public void updateUi(List<ItemAccount> items) {
        AccountChooserAdapter adapter = new AccountChooserAdapter(items, object -> {
            if (object != null) {
                try {
                    Intent intent = new Intent();
                    intent.putExtra(EXTRA_SELECTED_ITEM, new ObjectMapper().writeValueAsString(object));
                    intent.putExtra(EXTRA_SELECTED_OBJECT_TYPE, object.getClass().getName());
                    setResult(RESULT_OK, intent);
                    finish();
                } catch (JsonProcessingException e) {
                    throw new RuntimeException(e);
                }
            }
        });
        binding.recyclerview.setAdapter(adapter);
        binding.recyclerview.setLayoutManager(new LinearLayoutManager(this));
    }

    public static void startForResult(Fragment fragment,
                                      int requestCode,
                                      PaymentRequestType paymentRequestType,
                                      String title) {

        Intent starter = new Intent(fragment.getContext(), AccountChooserActivity.class);
        starter.putExtra(EXTRA_REQUEST_TYPE, paymentRequestType);
        starter.putExtra(EXTRA_REQUEST_CODE, requestCode);
        starter.putExtra(EXTRA_ACTIVITY_TITLE, title);
        fragment.startActivityForResult(starter, requestCode);
    }

    @Override
    protected AccountChooserPresenter createPresenter() {
        return accountChooserPresenter;
    }

    @Override
    protected AccountChooserView getView() {
        return this;
    }

}

File: app/src/main/java/piuk/blockchain/android/ui/chooser/AccountChooserActivity.kt
package piuk.blockchain.android.ui.chooser

import android.app.Activity
import android.content.Context
import android.content.Intent
import android.os.Bundle
import android.support.v4.app.Fragment
import android.support.v7.widget.LinearLayoutManager
import com.fasterxml.jackson.core.JsonProcessingException
import kotlinx.android.synthetic.main.activity_account_chooser.*
import kotlinx.android.synthetic.main.toolbar_general.*
import piuk.blockchain.android.BuildConfig
import piuk.blockchain.android.R
import piuk.blockchain.android.injection.Injector
import piuk.blockchain.android.ui.account.ItemAccount
import piuk.blockchain.android.ui.base.BaseMvpActivity
import piuk.blockchain.android.util.extensions.gone
import piuk.blockchain.android.util.extensions.toSerialisedString
import piuk.blockchain.android.util.extensions.visible
import piuk.blockchain.android.util.helperfunctions.consume
import piuk.blockchain.android.util.helperfunctions.unsafeLazy
import javax.inject.Inject

class AccountChooserActivity : BaseMvpActivity<AccountChooserView, AccountChooserPresenter>(),
    AccountChooserView {

    @Suppress("MemberVisibilityCanBePrivate")
    @Inject lateinit var accountChooserPresenter: AccountChooserPresenter

    override val isContactsEnabled: Boolean = BuildConfig.CONTACTS_ENABLED
    override val accountMode: AccountMode by unsafeLazy {
        intent.getSerializableExtra(EXTRA_CHOOSER_MODE) as AccountMode
    }

    init {
        Injector.getInstance().presenterComponent.inject(this)
    }

    override fun onCreate(savedInstanceState: Bundle?) {
        super.onCreate(savedInstanceState)
        setContentView(R.layout.activity_account_chooser)

        require(intent.hasExtra(EXTRA_CHOOSER_MODE)) { "Chooser mode must be passed to AccountChooserActivity" }
        require(intent.hasExtra(EXTRA_ACTIVITY_TITLE)) { "Title string must be passed to AccountChooserActivity" }

        val title = intent.getStringExtra(EXTRA_ACTIVITY_TITLE)
        setupToolbar(toolbar_general, title)

        setSupportActionBar(toolbar_general)
        supportActionBar?.run { setDisplayHomeAsUpEnabled(true) }

        onViewReady()
    }

    override fun showNoContacts() {
        recyclerview.gone()
        layout_no_contacts.visible()
    }

    override fun onSupportNavigateUp(): Boolean = consume {
        setResult(Activity.RESULT_CANCELED)
        onBackPressed()
    }

    override fun updateUi(items: List<ItemAccount>) {
        val adapter = AccountChooserAdapter(items) { any ->
            if (any != null) {
                try {
                    val intent = Intent().apply {
                        putExtra(EXTRA_SELECTED_ITEM, any.toSerialisedString())
                        putExtra(EXTRA_SELECTED_OBJECT_TYPE, any.javaClass.name)
                    }

                    setResult(Activity.RESULT_OK, intent)
                    finish()
                } catch (e: JsonProcessingException) {
                    throw RuntimeException(e)
                }
            }
        }
        recyclerview.apply {
            this.adapter = adapter
            layoutManager = LinearLayoutManager(this@AccountChooserActivity)
        }
    }

    override fun createPresenter() = accountChooserPresenter

    override fun getView() = this

    companion object {

        private const val EXTRA_REQUEST_CODE = "piuk.blockchain.android.EXTRA_REQUEST_CODE"
        private const val EXTRA_CHOOSER_MODE = "piuk.blockchain.android.EXTRA_CHOOSER_MODE"

        const val EXTRA_SELECTED_ITEM = "piuk.blockchain.android.EXTRA_SELECTED_ITEM"
        const val EXTRA_SELECTED_OBJECT_TYPE = "piuk.blockchain.android.EXTRA_SELECTED_OBJECT_TYPE"
        const val EXTRA_ACTIVITY_TITLE = "piuk.blockchain.android.EXTRA_ACTIVITY_TITLE"

        fun startForResult(
                fragment: Fragment,
                accountMode: AccountMode,
                requestCode: Int,
                title: String
        ) {
            val starter = createIntent(fragment.context!!, accountMode, requestCode, title)
            fragment.startActivityForResult(starter, requestCode)
        }

        fun startForResult(
                activity: Activity,
                accountMode: AccountMode,
                requestCode: Int,
                title: String
        ) {
            val starter = createIntent(activity, accountMode, requestCode, title)
            activity.startActivityForResult(starter, requestCode)
        }

        private fun createIntent(
                context: Context,
                accountMode: AccountMode,
                requestCode: Int,
                title: String
        ): Intent = Intent(context, AccountChooserActivity::class.java).apply {
            putExtra(EXTRA_CHOOSER_MODE, accountMode)
            putExtra(EXTRA_REQUEST_CODE, requestCode)
            putExtra(EXTRA_ACTIVITY_TITLE, title)
        }
    }

}

enum class AccountMode {

    // Show all accounts for ShapeShift, ie BTC & BCH HD accounts, Ether
    ShapeShift,
    // Show only the contacts list
    ContactsOnly,
    // Show all bitcoin accounts, including HD + legacy addresses
    Bitcoin,
    // Show bitcoin accounts + summarised legacy addresses + "All wallets"
    BitcoinSummary,
    // Show all bitcoin cash accounts, including HD + legacy addresses
    BitcoinCash,
    // Show bitcoin cash accounts + summarised legacy addresses + "All wallets"
    BitcoinCashSummary

}

